docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
+
+class ExplosiveTennisBall extends DataObject
+{
+    private static $has_one = [
+        'Thumbnail' => Image::class
+    ];
+
+    private static $has_many = [
+        'Tags' => BlogPost::class
+    ];
+
+
+    public function getShippingCost()
+    {
+        return Injector::inst(ProductService::class)->calculateCost($this);
+    }
+}
+```
+
+Alternatively, you can spell out the full qualified name of each class in a string. For example, `'Image'` would become `'SilverStripe\\Assets\\Image'`. Note the use of the _double backslash_ — this is necessary because the backslash is an escape character.
+
+#### Update references to classes in your YML config
+
+YAML configuration files can reference SilverStripe classes. Those references also need to use the fully qualified name of each class.
+
+Take for example the following SilverStripe 3 YAML configuration file.
+
+```yaml
+Injector:
+  ProductService:
+    properties:
+      RoadRunnerSpeed: 99999999
+      CoyoteSpeed: 1
+
+BlogPost:
+  extensions:
+    - HasOneExplosiveTennisBallExtension
+
+Email:
+  admin_email: no-reply@example.com
+```
+
+In SilverStripe 4, this will become:
+```yaml
+SilverStripe\Core\Injector\Injector:
+  Acme\Web\ProductService:
+    properties:
+      RoadRunnerSpeed: 99999999
+      CoyoteSpeed: 1
+
+SilverStripe\Blog\Model\BlogPost:
+  extensions:
+    - Acme\Web\Extensions\HasOneExplosiveTennisBallExtension
+
+SilverStripe\Control\Email\Email:
+  admin_email: no-reply@example.com
+```
+
+#### Update references to classes in your language files
+
+Translation keys are normally tied to classes. If you override SilverStripe's default translation or if you are localising your own project, you'll need to update those references to use the fully qualified name of each class.
+
+For example, let's say you had the following translation file in `mysite/lang/eng.yml`.
+```yaml
+en:
+  Member:
+    SINGULARNAME: Coyote
+  RoadRunner:
+    SALUTATION: Beep Beep
+```
+
+In SilverStripe 4, it would become:
+```yaml
+en:
+  SilverStripe\Security\Member:
+    SINGULARNAME: Coyote
+  Acme\Web\RoadRunner:
+    SALUTATION: Beep Beep
+```
+
+### Finalising namespace updates
+Whatever you upgraded your code automatically or manually you'll need to perform the two following steps manually.
+
+DataObject tables will default to use a namespaced name. For example, if you have class under `Acme\Web\Products\ExplosiveTennisBall` that extends `DataObject`, the matching table in your database will be called `Acme_Web_Products_ExplosiveTennisBall`.
+
+You can define a `private static $table_name` property on your DataObjects to use more convenient table names. For example, `private static $table_name = 'ExplosiveTennisBall';`.
+
+If your template files contain translatable strings, they also need to be updated to referenced the namespaced classes.
+
+For example, `<%t Member.SINGULARNAME 'Member' %>` would become `<%t SilverStripe\Security\Member.SINGULARNAME 'Member' %>`.
```

Comment:
You haven't mentioned here the fact that you can't use `static::class` or `parent::class` in when upgrading string representations in _t() usage.

I'd make it clear that in those cases either `__CLASS__` or `self::class` has to be used. I'd also mention this further up when talking about converting from strings to namespaces

Reply:
Reference: https://github.com/silverstripe/silverstripe-framework/issues/7647

Reply:
Done.